### PR TITLE
#169258642 Manager Accept Or  Reject  User Requests

### PR DESCRIPTION
**Pull request by william000000**
#### What does this PR do?
- Reject and accept Request (By Requester’s Manager)
#### Description of Task to be completed?
- Manager can approve his/her user trip requests
- Manager can reject his/her user trip requests
- Add validation to prevent unexpected issues
- Test the functionality of this API
- Add swagger documentation

#### How should this be manually tested?
- You have to clone the repo `git clone https://github.com/andela/team-odd-bn-backend.git`
- Then After the repo cloned `cd project-folder` and run `npm i` to install all dependencies
- Go to my branch `git checkout accept-and-reject-request-169258642`
- Run app by `npm run dev-start`
- Open `Postman` app
- Navigate to `localhost:{server_port}/api/v1/trips/{:tripRequestId}?status=accept` when approving a request or `localhost:{server_port}/api/v1/trips/{:tripRequestId}?status=reject` for rejecting a request
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
- [#169258642](https://www.pivotaltracker.com/story/show/169258642)
##### On success(approve)
image
##### On success(reject)
image
##### When Trip not exist 
image
##### When The Trip is not for that Manager
image
##### Swagger Documentation
image

#### Questions:
- N/A

### CODE REVIEW

File: src/routes/api/tripRoute.js [hunk 1]
```diff
@@ -239,4 +250,72 @@ tripRouter.post('/twoWay',
...
+/**
+ * @swagger
+ *
+ * /trips/{tripRequestId}?status=reject:
+ *    patch:
+ *      summary: Manager can reject a trip request
+ *      tags: [Trip]
+ *      requestBody:
+ *        required: true
+ *        content:
```
Comment (victkarangwa) [inline src/routes/api/tripRoute.js:263]: @william000000 I think docs come first then route
```
/**
*@swagger 
* ....
*/
tripRouter.post('/oneway',
  verifyToken,	  verifyToken, ...)
```

File: src/routes/api/tripRoute.js [hunk 2]
```diff
@@ -239,4 +250,72 @@ tripRouter.post('/twoWay',
  *
  */
 
+tripRouter.patch('/:tripRequestId', verifyToken, Validate.approveOrRejectRequest(), checkInputDataError, Exists.isTripRequestExist, isManagerRequest, isManagerAccess, validateAcceptOrReject, acceptOrRejectRequests);
+/**
```
Comment (victkarangwa) [inline src/routes/api/tripRoute.js:254]: @william000000, Are these two validations different: `Validate.approveOrRejectRequest()` and `validateAcceptOrReject`
I think they should be placed together

**Review by victkarangwa — CHANGES_REQUESTED** [review]

File: src/middlewares/approveOrReject.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  const findRequest = await tripRequests.findOne({
+    where: { userId: allUserId, id: tripRequestId }, raw: true
+  });
+
+  const { statusId } = findRequest;
+  const { status } = req.query;
+
+  if (status === 'reject') {
+    if (statusId === 3) {
+      return Customize.errorMessage(req, res, 'this request already rejected', 409);
```
Comment (victkarangwa) [inline src/middlewares/approveOrReject.js:28]: Just grammatical accuracy, I think it should be `This request is already rejected`

File: src/middlewares/approveOrReject.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    const rejectIt = await tripRequests.update({ statusId: 3 },
+      { where: { userId: findRequest.userId, id: tripRequestId } });
+    return !rejectIt[0] ? Customize.errorMessage(req, res, 'This request is not belongs to this Manager', 403) : next();
+  }
```
Comment (victkarangwa) [inline src/middlewares/approveOrReject.js:33]: The same here, `This request does not belong to this Manager`

**Review by victkarangwa — CHANGES_REQUESTED** [review]
Nice work @william000000
However, I left some feedback that we should work on.

File: src/tests/mock/mockData.js [hunk 5]
```diff
@@ -165,6 +165,17 @@ const mockData = {
     department: 'IT',
     managerId: 5,
     bio: 'I have been a christian since 2014, Christ as my savior'
+  },
+  managersUser: {
+    email: 'maiyo.brian@andela.com',
```
Comment (hezronkimutai) [inline src/tests/mock/mockData.js:170]: change this email to manager@gmail.com

File: src/routes/api/tripRoute.js (hunk 2 above)
Comment (william000000) [reply src/routes/api/tripRoute.js:254]: one is coming from the validation folder, another one is a middleware to operate the functionality of the task for this route .

**Pull request closed without merging**